### PR TITLE
Add DB2 support for Docker Compose

### PR DESCRIPTION
**Pull request by quaff**

### CODE REVIEW

**Comment by quaff** [comment]
`Db2R2dbcDockerComposeConnectionDetailsFactory` removed since `com.ibm.db2:db2-r2dbc` is not an actual driver.

```
Unable to create a ConnectionFactory for 'ConnectionFactoryOptions{options={database=testdb, host=127.0.0.1, driver=db2, password=REDACTED, port=50000, user=db2inst1}}'. Available drivers: [ oracle, sqlserver, postgresql ]
java.lang.IllegalStateException: Unable to create a ConnectionFactory for 'ConnectionFactoryOptions{options={database=testdb, host=127.0.0.1, driver=db2, password=REDACTED, port=50000, user=db2inst1}}'. Available drivers: [ oracle, sqlserver, postgresql ]
	at io.r2dbc.spi.ConnectionFactories.get(ConnectionFactories.java:143)
```

File: spring-boot-project/spring-boot-docker-compose/src/dockerTest/resources/org/springframework/boot/docker/compose/service/connection/db2/db2-compose.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    privileged: true
+    environment:
+      - 'LICENSE=accept'
+      - 'DB2INSTANCE=db2inst1'
+      - 'DB2INST1_PASSWORD=secret'
+      - 'DBNAME=testdb'
+      - 'AUTOCONFIG=false'
+      - 'ARCHIVE_LOGS=false'
+    healthcheck:
+      test: [ 'CMD', 'sleep', '115' ]
```
Comment (quaff) [inline spring-boot-project/spring-boot-docker-compose/src/dockerTest/resources/org/springframework/boot/docker/compose/service/connection/db2/db2-compose.yaml:15]: Didn't find better way for waiting DB2 fully up.

Comment (quaff) [reply spring-boot-project/spring-boot-docker-compose/src/dockerTest/resources/org/springframework/boot/docker/compose/service/connection/db2/db2-compose.yaml:15]: Testcontainers `Db2Container` use `LogMessageWaitStrategy() .withRegEx(".*Setup has completed\\..*")` , is there an equivalent way with compose file?

**Comment by quaff** [comment]
Need more disk space to run tests.

File: spring-boot-project/spring-boot-tools/spring-boot-test-support-docker/src/main/java/org/springframework/boot/testsupport/container/TestImage.java [hunk 2]
```diff
@@ -90,6 +92,12 @@ public enum TestImage {
 			(container) -> ((CouchbaseContainer) container).withStartupAttempts(5)
 				.withStartupTimeout(Duration.ofMinutes(10))),
 
+	/**
+	 * A container image suitable for testing DB2.
+	 */
+	DB2("ibmcom/db2", "11.5.8.0", () -> Db2Container.class,
```
Comment (mhalbritter) [inline spring-boot-project/spring-boot-tools/spring-boot-test-support-docker/src/main/java/org/springframework/boot/testsupport/container/TestImage.java:98]: https://hub.docker.com/r/ibmcom/db2 doesn't look like it's going to be updated in the future.

File: spring-boot-project/spring-boot-docker-compose/src/dockerTest/resources/org/springframework/boot/docker/compose/service/connection/db2/db2-compose.yaml [hunk 3]
```diff
@@ -0,0 +1,18 @@
+services:
+  database:
+    image: '{imageName}'
+    ports:
+      - '50000'
+    privileged: true
```
Comment (mhalbritter) [inline spring-boot-project/spring-boot-docker-compose/src/dockerTest/resources/org/springframework/boot/docker/compose/service/connection/db2/db2-compose.yaml:6]: I don't think you can run `privileged` containers on the CI.

File: spring-boot-project/spring-boot-docker-compose/src/dockerTest/resources/org/springframework/boot/docker/compose/service/connection/db2/db2-compose.yaml [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+      - 'DB2INSTANCE=db2inst1'
+      - 'DB2INST1_PASSWORD=secret'
+      - 'DBNAME=testdb'
+      - 'AUTOCONFIG=false'
+      - 'ARCHIVE_LOGS=false'
+    healthcheck:
+      test: 'su - db2inst1 -c "db2 connect to testdb user db2inst1 using secret"'
+      start_period: 90s
+    labels:
+      org.springframework.boot.readiness-check.tcp.disable: true
```
Comment (mhalbritter) [inline spring-boot-project/spring-boot-docker-compose/src/dockerTest/resources/org/springframework/boot/docker/compose/service/connection/db2/db2-compose.yaml:18]: Isn't this using TCP?

**Review by mhalbritter — COMMENTED** [review]
Thanks for the PR, @quaff . I've left some comments for your consideration.

File: spring-boot-project/spring-boot-tools/spring-boot-test-support-docker/src/main/java/org/springframework/boot/testsupport/container/TestImage.java (hunk 2 above)
Comment (quaff) [reply spring-boot-project/spring-boot-tools/spring-boot-test-support-docker/src/main/java/org/springframework/boot/testsupport/container/TestImage.java:98]: The successor `icr.io/db2_community/db2` is supported also, but there is no page list tags, using `latest` is not a good option.

File: spring-boot-project/spring-boot-docker-compose/src/dockerTest/resources/org/springframework/boot/docker/compose/service/connection/db2/db2-compose.yaml (hunk 3 above)
Comment (quaff) [reply spring-boot-project/spring-boot-docker-compose/src/dockerTest/resources/org/springframework/boot/docker/compose/service/connection/db2/db2-compose.yaml:6]: Otherwise `Connection reset` exception always be thrown, that's why testcontainers `Db2Container` need that also.
https://github.com/testcontainers/testcontainers-java/blob/04206d9dff440e875906f01151f09ad04da3c68f/modules/db2/src/main/java/org/testcontainers/containers/Db2Container.java#L60

File: spring-boot-project/spring-boot-docker-compose/src/dockerTest/resources/org/springframework/boot/docker/compose/service/connection/db2/db2-compose.yaml (hunk 4 above)
Comment (quaff) [reply spring-boot-project/spring-boot-docker-compose/src/dockerTest/resources/org/springframework/boot/docker/compose/service/connection/db2/db2-compose.yaml:18]: It open tcp port prematurely, the database is not fully up.

File: spring-boot-project/spring-boot-docker-compose/build.gradle [hunk 5]
```diff
@@ -23,6 +23,7 @@ dependencies {
 	dockerTestRuntimeOnly("io.r2dbc:r2dbc-mssql")
 	dockerTestRuntimeOnly("org.postgresql:postgresql")
 	dockerTestRuntimeOnly("org.postgresql:r2dbc-postgresql")
+	dockerTestRuntimeOnly("com.ibm.db2:jcc")
```
Comment (snicoll) [inline spring-boot-project/spring-boot-docker-compose/build.gradle:26]: Dependencies are meant to be ordered alphabetically. Please move this at the top of the group.

File: spring-boot-project/spring-boot-tools/spring-boot-test-support-docker/src/main/java/org/springframework/boot/testsupport/container/TestImage.java [hunk 6]
```diff
@@ -84,14 +86,20 @@ public enum TestImage {
 			(container) -> ((CassandraContainer<?>) container).withStartupTimeout(Duration.ofMinutes(10))),
 
 	/**
-	 * A container image suitable for testing Couchbase.
+	 * A Docker image suitable for running.
```
Comment (snicoll) [inline spring-boot-project/spring-boot-tools/spring-boot-test-support-docker/src/main/java/org/springframework/boot/testsupport/container/TestImage.java:89]: Uh?

File: spring-boot-project/spring-boot-tools/spring-boot-test-support-docker/build.gradle [hunk 7]
```diff
@@ -15,6 +15,7 @@ dependencies {
 	compileOnly("org.springframework:spring-core")
 
 	optional("org.testcontainers:activemq")
+	optional("org.testcontainers:db2")
```
Comment (snicoll) [inline spring-boot-project/spring-boot-tools/spring-boot-test-support-docker/build.gradle:18]: This is also ordered alphabetically.

File: spring-boot-project/spring-boot-docker-compose/src/dockerTest/resources/org/springframework/boot/docker/compose/service/connection/db2/db2-compose.yaml [hunk 8]
```diff
@@ -0,0 +1,18 @@
+services:
+  database:
+    image: '{imageName}'
+    ports:
+      - '50000'
+    privileged: true
+    environment:
+      - 'LICENSE=accept'
```
Comment (snicoll) [inline spring-boot-project/spring-boot-docker-compose/src/dockerTest/resources/org/springframework/boot/docker/compose/service/connection/db2/db2-compose.yaml:8]: What does this mean in practice?

File: spring-boot-project/spring-boot-tools/spring-boot-test-support-docker/src/main/java/org/springframework/boot/testsupport/container/TestImage.java [hunk 9]
```diff
@@ -130,7 +138,7 @@ public enum TestImage {
 	MARIADB("mariadb", "10.10"),
 
 	/**
-	 * A container image suitable for testing MongoDB.
+	 * A Docker image suitable for MongoDB.
```
Comment (snicoll) [inline spring-boot-project/spring-boot-tools/spring-boot-test-support-docker/src/main/java/org/springframework/boot/testsupport/container/TestImage.java:141]: This is unrelated, please revert.

**Review by snicoll — CHANGES_REQUESTED** [review]
I've added some comment, please have a look when you get a chance.

File: spring-boot-project/spring-boot-docker-compose/src/dockerTest/resources/org/springframework/boot/docker/compose/service/connection/db2/db2-compose.yaml (hunk 8 above)
Comment (quaff) [reply spring-boot-project/spring-boot-docker-compose/src/dockerTest/resources/org/springframework/boot/docker/compose/service/connection/db2/db2-compose.yaml:8]: Which part?

File: spring-boot-project/spring-boot-tools/spring-boot-test-support-docker/src/main/java/org/springframework/boot/testsupport/container/TestImage.java (hunk 2 above)
Comment (quaff) [reply spring-boot-project/spring-boot-tools/spring-boot-test-support-docker/src/main/java/org/springframework/boot/testsupport/container/TestImage.java]: OK, I changed it to `icr.io/db2_community/db2`

File: spring-boot-project/spring-boot-docker-compose/src/dockerTest/resources/org/springframework/boot/docker/compose/service/connection/db2/db2-compose.yaml (hunk 8 above)
Comment (snicoll) [reply spring-boot-project/spring-boot-docker-compose/src/dockerTest/resources/org/springframework/boot/docker/compose/service/connection/db2/db2-compose.yaml:8]: The `LICENSE=accept` bit.

File: spring-boot-project/spring-boot-docs/src/docs/antora/modules/appendix/pages/application-properties/index.adoc [hunk 10]
```diff
@@ -11,39 +11,38 @@ Make sure to review xref:reference:features/external-config.adoc#features.extern
 NOTE: Property contributions can come from additional jar files on your classpath, so you should not consider this an exhaustive list.
 Also, you can define your own properties.
 
-
-include::partial$configuration-properties/core.adoc[]
+include::partial$configuration-properties/actuator.adoc[]
```
Comment (snicoll) [inline spring-boot-project/spring-boot-docs/src/docs/antora/modules/appendix/pages/application-properties/index.adoc:14]: This commit shouldn't be part of the PR.

Comment (quaff) [reply spring-boot-project/spring-boot-docs/src/docs/antora/modules/appendix/pages/application-properties/index.adoc]: My mistake, I dropped it.

File: spring-boot-project/spring-boot-docker-compose/src/dockerTest/resources/org/springframework/boot/docker/compose/service/connection/db2/db2-compose.yaml (hunk 8 above)
Comment (quaff) [reply spring-boot-project/spring-boot-docker-compose/src/dockerTest/resources/org/springframework/boot/docker/compose/service/connection/db2/db2-compose.yaml:8]: It's required by starting DB2 instance, testcontainers' `DB2Container` need this also. see https://github.com/testcontainers/testcontainers-java/blob/04206d9dff440e875906f01151f09ad04da3c68f/modules/db2/src/main/java/org/testcontainers/containers/Db2Container.java#L104

Comment (snicoll) [reply spring-boot-project/spring-boot-docker-compose/src/dockerTest/resources/org/springframework/boot/docker/compose/service/connection/db2/db2-compose.yaml:8]: I understand that it's required, I am asking what it means in practice that our project starts accepting that license.

Comment (quaff) [reply spring-boot-project/spring-boot-docker-compose/src/dockerTest/resources/org/springframework/boot/docker/compose/service/connection/db2/db2-compose.yaml:8]: Not for production use, I guess.

Comment (wilkinsona) [reply spring-boot-project/spring-boot-docker-compose/src/dockerTest/resources/org/springframework/boot/docker/compose/service/connection/db2/db2-compose.yaml:8]: I don't think this is resolved as we haven't determined if we can legally accept the license for testing purposes. Knowing that Testcontainers has support for accepting the license is useful but it's opt-in and doesn't tell us if we can do the same automatically on CI and dev machines for testing purposes.

Comment (quaff) [reply spring-boot-project/spring-boot-docker-compose/src/dockerTest/resources/org/springframework/boot/docker/compose/service/connection/db2/db2-compose.yaml:8]: It's not opt-in, Testcontainers accept the license by default.

Comment (eddumelendez) [reply spring-boot-project/spring-boot-docker-compose/src/dockerTest/resources/org/springframework/boot/docker/compose/service/connection/db2/db2-compose.yaml:8]: License is not accepted by default in Testcontainers. User must call `acceptLicense()`, which sets the env var, when defining `Db2Container`. See https://github.com/testcontainers/testcontainers-java/blob/main/modules/db2/src/test/java/org/testcontainers/junit/db2/SimpleDb2Test.java#L17

Comment (quaff) [reply spring-boot-project/spring-boot-docker-compose/src/dockerTest/resources/org/springframework/boot/docker/compose/service/connection/db2/db2-compose.yaml:8]: My mistake, I forgot I added `container-license-acceptance.txt`.

**Comment by wilkinsona** [comment]
I've force-pushed this having rebased it on the latest main. That should mean the PR picks up some recent changes to the runner preparation that will make more free disk space available.

**Comment by wilkinsona** [comment]
This is blocked until the questions regarding the licence acceptance have been resolved.

**Comment by philwebb** [comment]
I looks like DB2 is also being removed from Docker Hub (see https://community.ibm.com/community/user/datamanagement/blogs/vijaya-katikireddy/2023/02/04/db2-community). Given that and the license issues, I'm not sure that we should try to support this.

**Comment by quaff** [comment]
> I looks like DB2 is also being removed from Docker Hub (see https://community.ibm.com/community/user/datamanagement/blogs/vijaya-katikireddy/2023/02/04/db2-community). Given that and the license issues, I'm not sure that we should try to support this.

This PR use new `icr.io/db2_community/db2` for test.

**Comment by wilkinsona** [comment]
I believe that [this is the relevant license](https://www.ibm.com/support/customer/csol/terms/?id=L-PGUL-3Z3FFG&lc=en).

It contains the following usage restriction:

```
1. Usage Restrictions

(a) The Program may use a maximum of 4 processor cores and 16 GB of memory on each physical or virtual server.
```

I think we could comply with the license by setting some [CPU](https://docs.docker.com/reference/compose-file/deploy/#cpus) and [memory](https://docs.docker.com/reference/compose-file/deploy/#memory) limits in the compose files used for testing. Whether or not it's worth the trouble is another question.

**Comment by philwebb** [comment]
I think unfortunately the usage restrictions, combined with the fact that the image isn't in Docker Hub means we probably shouldn't add support for this. Thanks anyway for the PR @quaff!
